### PR TITLE
chore: add template for new generators

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR adds a template that can be used to easily setup new generators with the bare minimum.

**Related issue(s)**
Fixes https://github.com/asyncapi/modelina/issues/256

### CODE REVIEW

**Comment by jonaslagoni** [comment]
@leigh-johnson as you are the one who most recently added a new generator, if you have the time, I would love to know if you find this useful 🙂 

I am gonna try to go through it my self and add the python generator to validate nothing is missing from my side, but would be great to have fresh pair of eyes 😄

File: docs/contributing.md [hunk 1]
```diff
@@ -35,6 +35,38 @@ Adding examples is quite straight forward, so don't feel shy! Here's how to do i
...
+3. Replace the filenames `Template...` with your generator name.
+4. Add your [generator to the generator index file](../src/generators/index.ts).
+
+Now it's time to adapt the template into what ever it is you are generating:
+1. Adapt the [constraint logic](../src/generators/template/) based on what is allowed within your output. Read more about the constraint logic here.
+2. Add all of the reserved keywords that the models must never generate in the [Constant file](../src/generators/template/Constants.ts).
+3. Adapt/create the first renderers. The template by default include two renderers, one for rendering enums and one for classes, but you can create what ever renderers makes sense. For example in Rust it's not called class but struct, so therefore its called a `StructRenderer`.
+4. Adapt the file generator and the rendering of the complete models to fit your generator.
+
+Time to adapt the tests, cause without tests, it's just an empty promise. The test that is included in the template is really just placeholders, so make sure you adapt them accordingly to your code.
```
Comment (leigh-johnson) [inline docs/contributing.md:55]: A note about black box tests might be helpful here! These tests were the most useful in finding corner cases.

File: docs/contributing.md [hunk 2]
```diff
@@ -35,6 +35,38 @@ Adding examples is quite straight forward, so don't feel shy! Here's how to do i
...
+4. Add your generator to the [FileGenerators test](../test/generators/FileGenerators.spec.ts) to ensure the models are accurately written to files.
+
+Lastly, we need to adapt some of the docs to showcase your new awesome generator! Cause if the users cant find it, it dont exist.
+1. Add your [generator specific documentation under languages](./languages/) and add it to the [list of generators](./README.md#languages)
+2. Add your generator to the list of generators in the [main readme file](../README.md)
+3. Add a basic [usage example to the usage documentation](./usage.md), you can see more about how to create [examples here](#adding-examples).
+
+Aaaand that's it! As a rule of thumb, start small and slowly add more features, dont try to push everything into one PR, as it will take forever to review, code, and merge. 
+
+PR's you can look to for guidance on how the process goes: https://github.com/asyncapi/modelina/pull/818
```
Comment (leigh-johnson) [inline docs/contributing.md:68]: I wonder if Presets deserve a section here? When I was diving in, I had a hard time figuring out what to put in a "base" Preset and the interplay with Preset overrides. Going through the Java generator helped me understand the abstraction level, so that's a great example to link!

File: docs/contributing.md [hunk 3]
```diff
@@ -35,6 +35,38 @@ Adding examples is quite straight forward, so don't feel shy! Here's how to do i
 
 Aaaand you are done! :tada: 
 
+### New Generators
+Generators sits as the core of Modelina, which frames the core concepts of what you can generate. Therefore it's also no small task to create a new one, so dont get discourage, we are here to help you! 
+
+To make it easier to contribute a new generator, and to avoid focusing too much of the internals of Modelina, we created a template generator to get you started. If you encounter discreprencies with the following guide or templates, make sure to raise it as an issue so it can be fixed!
+
```
Comment (leigh-johnson) [inline docs/contributing.md:42]: I'd throw a link into Slack here, to encourage potential contributors to join/chat before diving into their implementation. I had to rifle around looking for IRC/chat channel.

File: docs/contributing.md (hunk 1 above)
Comment (jonaslagoni) [reply docs/contributing.md:55]: You are right! 

Also helped me run down a few bugs 😆

File: docs/contributing.md (hunk 2 above)
Comment (jonaslagoni) [reply docs/contributing.md:68]: That makes sense, adding that 👍

File: docs/contributing.md (hunk 3 above)
Comment (jonaslagoni) [reply docs/contributing.md:42]: Agree 👍

**Comment by jonaslagoni** [comment]
> 👍 Very nice work, thank you for putting together this template! Couple of quick notes that sprung to mind based on my recent experience writing the RustGenerator. I block out some time to put 👀 on the Python generator next week if you're looking for feedback.

Thanks for the feedback!

Would definitely appreciate some eyes on the generator if you have time! Cause my python is non-existing, I can create the generation, but knowing if it's the right output is another talk 😅 You can find the PR here: https://github.com/asyncapi/modelina/pull/863

File: docs/contributing.md (hunk 1 above)
Comment (jonaslagoni) [reply docs/contributing.md:55]: I ended up adding the following:

```
6. Lastly, we have (arguably) the most important tests, [the BlackBox tests](./development.md#blackbox-testing). They are to ensure that real-world inputs generate usable models that do not contain syntax errors. You can read more about the BlackBox tests [here](./development.md#blackbox-testing).
```

We can then use `development.md#blackbox-testing` to add more details about black-box testing if needed. I hope that suffice 🤞

File: docs/contributing.md (hunk 2 above)
Comment (jonaslagoni) [reply docs/contributing.md:68]: I added the following:

```
An important note about presets, they are used to extend and build upon the default model, the bare minimum of a data model, so that Modelina can support multiple features. You can read more about [presets here](./presets.md). If you have any questions or want something clarified, don't hesitate to [reach out on slack](https://asyncapi.com/slack-invite).
```

Let me know if we need to add anything else 🤔

File: docs/contributing.md (hunk 3 above)
Comment (jonaslagoni) [reply docs/contributing.md:42]: I ended up adding the following
```
If you have any questions, are unsure how your use-case fits in, or want something clarified, don't hesitate to [reach out on slack](https://asyncapi.com/slack-invite).
```

**Review by leigh-johnson — DISMISSED** [review]
👀 The external link validation test failure looks like an intermittent flake. This link works for me!

```
ERROR: 1 dead links found!
[✖] <https://img.shields.io/github/workflow/status/asyncapi/modelina/Blackbox%20testing%20(Stay%20Awhile%20and%20Listen)?label=blackbox%20testing> → Status: 400
````

File: docs/contributing.md [hunk 4]
```diff
@@ -35,6 +35,47 @@ Adding examples is quite straight forward, so don't feel shy! Here's how to do i
...
+#### Getting started
+1. Start by copy/pasting the [template generator](../src/generators/template/) and [tests](../test/generators/template/) and rename it to your generator.
+2. Search and replace within your new generator and test folder for `Template`, `template` and `TEMPLATE` and replace it with your generator name and match the cases. **Make sure you search and replace it with matching case**.
+3. Replace the filenames `Template...` with your generator name.
+4. Add your [generator to the generator index file](../src/generators/index.ts).
+
+Now it's time to adapt the template into what ever it is you are generating:
+1. Adapt the [constraint logic](../src/generators/template/constrainer) and [the type constraints](../src/generators/template/TemplateConstrainer.ts) based on what is allowed within your output. [Read more about the constraint logic here](./constraints.md).
```
Comment (magicmatatjahu) [inline docs/contributing.md:50]:
```suggestion
1. Adapt the [constraint logic](../src/generators/template/constrainer) and the [type constraints](../src/generators/template/TemplateConstrainer.ts) based on what is allowed within your output. Read more about the constraint logic [here](./constraints.md).
```

File: docs/contributing.md [hunk 5]
```diff
@@ -35,6 +35,47 @@ Adding examples is quite straight forward, so don't feel shy! Here's how to do i
...
+7. Lastly, we have (arguably) the most important tests, [the BlackBox tests](./development.md#blackbox-testing). They are to ensure that real-world inputs generate usable models that do not contain syntax errors. You can read more about the BlackBox tests [here](./development.md#blackbox-testing).
+
+Lastly, we need to adapt some of the docs to showcase your new awesome generator! Cause if the users cant find it, it dont exist.
+1. Add your [generator specific documentation under languages](./languages/) and add it to the [list of generators](./README.md#languages)
+2. Add your generator to the list of generators in the [main readme file](../README.md)
+3. Add a basic [usage example to the usage documentation](./usage.md), you can see more about how to create [examples here](#adding-examples).
+
+Aaaand that's it! As a rule of thumb, start small and slowly add more features, dont try to push everything into one PR, as it will take forever to review, code, and merge. 
+
+If you have any questions, are unsure how your use-case fits in, or want something clarified, don't hesitate to [reach out on slack](https://asyncapi.com/slack-invite).
```
Comment (magicmatatjahu) [inline docs/contributing.md:73]: We should write that before `#### Getting started` section.

File: docs/contributing.md [hunk 6]
```diff
@@ -35,6 +35,47 @@ Adding examples is quite straight forward, so don't feel shy! Here's how to do i
...
+6. Add your generator to the [FileGenerators test](../test/generators/FileGenerators.spec.ts) to ensure the models are accurately written to files.
+7. Lastly, we have (arguably) the most important tests, [the BlackBox tests](./development.md#blackbox-testing). They are to ensure that real-world inputs generate usable models that do not contain syntax errors. You can read more about the BlackBox tests [here](./development.md#blackbox-testing).
+
+Lastly, we need to adapt some of the docs to showcase your new awesome generator! Cause if the users cant find it, it dont exist.
+1. Add your [generator specific documentation under languages](./languages/) and add it to the [list of generators](./README.md#languages)
+2. Add your generator to the list of generators in the [main readme file](../README.md)
+3. Add a basic [usage example to the usage documentation](./usage.md), you can see more about how to create [examples here](#adding-examples).
+
+Aaaand that's it! As a rule of thumb, start small and slowly add more features, dont try to push everything into one PR, as it will take forever to review, code, and merge.
```
Comment (magicmatatjahu) [inline docs/contributing.md:71]:
```suggestion
Aaaand that's it! As a rule of thumb, start small and slowly add more features, don't try to push everything into one PR, as it will take forever to review, code, and merge. 
```

File: src/generators/template/TemplateGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    const result = await renderer.runSelfPreset();
+    return RenderOutput.toRenderOutput({result, renderedName: model.name, dependencies: renderer.dependencies});
+  }
+
+  async renderEnum(model: ConstrainedEnumModel, inputModel: InputMetaModel): Promise<RenderOutput> {
+    const presets = this.getPresets('enum'); 
+    const renderer = new EnumRenderer(this.options, this, presets, model, inputModel);
+    const result = await renderer.runSelfPreset();
+    return RenderOutput.toRenderOutput({result, renderedName: model.name, dependencies: renderer.dependencies});
+  }
```
Comment (magicmatatjahu) [inline src/generators/template/TemplateGenerator.ts:115]: Hmm, one renderer type like class is enough for me. Not every language has something like enum but almost every has class or struct.

**Review by magicmatatjahu — COMMENTED** [review]
Awesome! I have some comments, please refer to them :)

File: src/generators/template/TemplateGenerator.ts (hunk 7 above)
Comment (jonaslagoni) [reply src/generators/template/TemplateGenerator.ts:115]: It is not as much for the enum's sake, but to have two renderers, whether it's for the enum or something else does not really matter 🙂 So I would say it's best just to keep it, once it has been used a bit we can always bring up the discussion whether it should be removed.

File: docs/contributing.md (hunk 5 above)
Comment (jonaslagoni) [reply docs/contributing.md:73]: Moved it to the beginning of the document, I think it made more sense there as it points out where to get help right from the start 🙂

**Review by magicmatatjahu — APPROVED** [review]
LGTM!
